File: clientes/EventosClientes.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

"""
    Nombre: EventosClientes.py
    Autor: Lic. Claudio Invernizzi
    Fecha: 18/10/25
    E-Mail: cinvernizzi@dsgestion.site
    Proyecto: Presupuesto
    Comentarios: Clase que controla las operaciones sobre el 
                 formulario de clientes

"""

# importamos las librerías
from datetime import datetime
from clientes.Clientes import Clientes
from clases.mensaje import Mensaje

class EventosClientes:
    """
    
    :author: Claudio Invernizzi <cinvernizzi@dsgestion.site>

    Definición de la clase 

    """

    def __init__(self, formclientes):
        """

        :author: Claudio Invernizzi <cinvernizzi@dsgestion.site>

        :param: formclientes el formulario de datos de los clientes

        Constructor de la clase, recibe como parámetro el formulario 
        de clientes

        """

        # asignamos en la clase
        self.FormClientes = formclientes

        # instanciamos la clase de la base de datos 
        self.Clientes = Clientes()

    def nuevoCliente(self):
        """

        :author: Claudio Invernizzi <cinvernizzi@dsgestion.site>

        Método que limpia el formulario preparándolo para un nuevo 
        ingreso

        """

        # limpiamos
        self.FormClientes.tId.clear()
        self.FormClientes.tNombre.clear()
        self.FormClientes.tDireccion.clear()
        self.FormClientes.tTributaria.clear()
        self.FormClientes.tTelefono.clear()
        self.FormClientes.tMail.clear()

        # fijamos la fecha de alta
        ahora = datetime.now()
        self.FormClientes.tAlta.setText(ahora.strftime("%d/%m/%Y"))        

        # fijamos el foco
        self.FormClientes.tNombre.setFocus()

    def getDatosCliente(self, idcliente):
        """

        :author: Claudio Invernizzi <cinvernizzi@dsgestion.site>

        :param self: la propia clase
        :param idcliente: entero con la clave del cliente

        Método que recibe como parámetro la clave del cliente y obtiene
        los datos del mismo y los presenta en el formulario

        """

        # obtenemos el registro
        self.Clientes.getDatosCliente(idcliente)

        # asignamos en el formulario
        self.FormClientes.tId.setText(str(self.Clientes.Id))
        self.FormClientes.tNombre.setText(self.Clientes.Nombre)
        self.FormClientes.tDireccion.setText(self.Clientes.Direccion)
        self.FormClientes.tTributaria.setText(self.Clientes.Identificacion)
        self.FormClientes.tTelefono.setText(self.Clientes.Telefono)
        self.FormClientes.tMail.setText(self.Clientes.Mail)
        self.FormClientes.tAlta.setText(self.Clientes.Fecha)        

        # fijamos el foco
        self.FormClientes.tNombre.setFocus()

    def cancelaCliente(self):
        """
        
        :author: Claudio Invernizzi <cinvernizzi@dsgestion.site>
        
        Método llamado al pulsar sobre el botón cancelar que 
        según el estado recarga el registro o lo limpia 

        """

        # si está editando
        if self.FormClientes.tId.text():

            # recargamos el registro
            self.getDatosCliente(self.FormClientes.tId.text())

        # si está insertando
        else:

            # limpiamos el formulario
            self.nuevoCliente()

    def verificaCliente(self):
        """
        
        :author: Claudio Invernizzi <cinvernizzi@dsgestion.site>

        Método llamado al pulsar el botón grabar que verifica 
        los datos del formulario 

        """

        # si está dando de alta
        if not self.FormClientes.tId.text():
            self.Clientes.Id = 0
        else:
            self.Clientes.Id = self.FormClientes.tId.text()

        # verifica el nombre 
        if not self.FormClientes.tNombre.text():
            Mensaje(self.FormClientes, "Ingrese el nombre del cliente")
            return
        else:
            self.Clientes.Nombre = self.FormClientes.tNombre.text()

        # verifica el domicilio 
        if not self.FormClientes.tDireccion.text():
            Mensaje(self.FormClientes, "Indique el domicilio del cliente")
            return
        else:
            self.Clientes.Direccion = self.FormClientes.tDireccion.text()

        # verifica la clave tributaria
        if not self.FormClientes.tTributaria.text():
            Mensaje(self.FormClientes, "Indique la clave tributaria")
            return
        else:
            self.Clientes.Identificacion = self.FormClientes.tTributaria.text()

        # verifica que al menos exista teléfono o mail 
        if not self.FormClientes.tTelefono.text() and not self.FormClientes.tMail.text():
            Mensaje(self.FormClientes, "Indique al menos una forma de contacto")
            return
        else:
            self.Clientes.Telefono = self.FormClientes.tTelefono.text()
            self.Clientes.Mail = self.FormClientes.tMail.text()

        # si llegó hasta aquí grabamos el registro
        if int(self.Clientes.grabaCliente()) != 0:

            # si estaba insertando actualiza la fecha
            if not self.FormClientes.tAlta.text():
                ahora = datetime.now()
                self.FormClientes.tAlta.setText(ahora.strftime("%d/%m/%Y"))

            # asignamos en el formulario 
            self.FormClientes.tId.setText(str(id))
            Mensaje(self.FormClientes, "Registro grabado ... ")
        
        else:

            # presenta el mensaje
            Mensaje(self.FormClientes, "Ha ocurrido un error")

File: src/main/java/site/dsgestion/clientes/EventosClientes.java
/**
 * 
 * @author Claudio Invernizzi <cinvernizzi@dsgestion.site>
 * @date 23/01/2026
 * @name EventosClientes
 * @Projecto: Presupuesto
 * @Copyright DsGestion <dsgestion.site>
 * @Licence: GPL
 * Clase que controla los eventos del formulario de clientes
 * 
 */

// definimos el paquete
package site.dsgestion.clientes;

// importamos las librerías 
import site.dsgestion.dbApi.Utilidades;
import site.dsgestion.dbApi.Mensaje;

/**
 *
 * @author Claudio Invernizzi <cinvernizzi@dsgestion.site>
 * 
 * Definición de la clase
 * 
 */
public class EventosClientes {
   
    // declaramos las variables
    protected DbClientes Clientes;      // controlador de la base de datos
    protected FormClientes Formulario;  // formulario de datos
    protected Utilidades Herramientas;  // utilidades de fechas
    
    /**
     * 
     * @author Claudio Invernizzi <cinvernizzi@dsgestion.site>
     * 
     * @param formulario la clase con la definición del formulario
     * 
     * Constructor de la clase, recibe como parámetro el formulario 
     * con los datos 
     * 
     */
    public EventosClientes(FormClientes formulario){
        
        // asignamos en la clase
        this.Formulario = formulario;
        
        // instanciamos la clase de herramientes y de base de datos
        this.Herramientas = new Utilidades();
        this.Clientes = new DbClientes();
        
    }
    
    /**
     * 
     * @author Claudio Invernizzi <cinvernizzi@dsgestion.site>
     * 
     * Método llamado al pulsar el botón grabar que verifica 
     * los datos del formulario y en todo caso graba el registro
     * 
     */
    public void verificaCliente(){
        
        // si está insertando
        if (this.Formulario.tId.getText().equals("")){
            this.Clientes.setId(0);
        } else {
            this.Clientes.setId(Integer.parseInt(this.Formulario.tId.getText()));
        }
        
        // si no ingresó el nombre
        if (this.Formulario.tNombre.getText().equals("")){
            new Mensaje("Ingrese el nombre del cliente");
            return;
        } else {
            this.Clientes.setNombre(this.Formulario.tNombre.getText());
        }
     
        // si no ingresó la dirección
        if (this.Formulario.tDireccion.getText().equals("")){
            new Mensaje("Indique la dirección del cliente");
            return;
        } else {
            this.Clientes.setDireccion(this.Formulario.tDireccion.getText());
        }
        
        // si no ingresó la id tributaria
        if (this.Formulario.tIdTributaria.getText().equals("")){
            new Mensaje("No hay Id Tributaria, se graba igualmente");
        }
        this.Clientes.setIdentificacion(this.Formulario.tIdTributaria.getText());
        
        // si no ingresó ni teléfono ni mail
        if (this.Formulario.tTelefono.getText().equals("") &&
            this.Formulario.tMail.getText().equals("")){
            
            // presenta el mensaje y retorna
            new Mensaje("Debe indicar una forma de contacto");
            return;
            
        }
        
        // si ingresó el mail verificamos que sea correcto
        if (!this.Formulario.tMail.getText().equals("")){
            
            // verificamos 
            if (!this.Herramientas.esEmailCorrecto(this.Formulario.tMail.getText())){
                
                // presenta el mensaje
                new Mensaje("El mail parece incorrecto");
                return;
            }
            
        }
        
        // fijamos el teléfono y el mail
        this.Clientes.setTelefono(this.Formulario.tTelefono.getText());
        this.Clientes.setMail(this.Formulario.tMail.getText());
        
        // si llegó hasta aquí grabamos
        int id = this.Clientes.grabaCliente();
        
        // si salió todo bien
        if (id != 0){
        
            // fijamos la clave
            this.Formulario.tId.setText(Integer.toHexString(id));
            
            // presenta el mensaje
            new Mensaje("Registro grabado ...");
            
        }
        
    }
    
    /**
     * 
     * @author Claudio Invernizzi <cinvernizzi@dsgestion.site>
     * 
     * Método llamado al pulsar el botón cancelar que según el 
     * estado del formulario lo limpia o recarga el registro
     * 
     */
    public void cancelaCliente(){
        
        // si está insertando
        if (this.Formulario.tId.getText().equals("")){
            this.nuevoCliente();
        } else {
            this.getDatosCliente(Integer.parseInt(this.Formulario.tId.getText()));
        }
        
    }
    
    /**
     * 
     * @author Claudio Invernizzi <cinvernizzi@dsgestion.site>
     * 
     * Método que limpia el formulario de datos 
     * 
     */
    public void nuevoCliente(){
        
        // limpiamos el formulario
        this.Formulario.tId.setText("");
        this.Formulario.tNombre.setText("");
        this.Formulario.tDireccion.setText("");
        this.Formulario.tIdTributaria.setText("");
        this.Formulario.tTelefono.setText("");
        this.Formulario.tMail.setText("");
        this.Formulario.tFecha.setText(this.Herramientas.FechaActual());
        
        // fijamos el foco
        this.Formulario.tNombre.requestFocus();
        
    }
    
    /**
     * 
     * @author Claudio Invernizzi <cinvernizzi@dsgestion.site>
     * 
     * @param idcliente - entero con la clave del registro
     * 
     * Método que recibe como parámetro la clave de un registro 
     * y carga los datos del mismo
     * 
     */
    public void getDatosCliente(int idcliente){

        // obtenemos el registro
        this.Clientes.getDatosCliente(idcliente);
        
        // lo presentamos
        this.Formulario.tId.setText(Integer.toString(this.Clientes.getId()));
        this.Formulario.tNombre.setText(this.Clientes.getNombre());
        this.Formulario.tDireccion.setText(this.Clientes.getDireccion());
        this.Formulario.tIdTributaria.setText(this.Clientes.getIdentificacion());
        this.Formulario.tTelefono.setText(this.Clientes.getTelefono());
        this.Formulario.tMail.setText(this.Clientes.getMail());
        this.Formulario.tFecha.setText(this.Clientes.getFecha());
        
        // fijamos el foco
        this.Formulario.tNombre.requestFocus();
        
    }
    
}
